airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java

```diff
@@ -144,30 +144,39 @@ public void decrement(final long stateId, final long count) {
   public List<PartialAirbyteMessage> flushStates() {
     final List<PartialAirbyteMessage> output = new ArrayList<>();
     Long bytesFlushed = 0L;
-    for (final Map.Entry<StreamDescriptor, LinkedList<Long>> entry : streamToStateIdQ.entrySet()) {
-      // Remove all states with 0 counters.
-      // Per-stream synchronized is required to make sure the state (at the head of the queue)
-      // logic is applied to is the state actually removed.
-      synchronized (this) {
+    synchronized (this) {
+      for (final Map.Entry<StreamDescriptor, LinkedList<Long>> entry : descToStateIdQ.entrySet()) {
+        // Remove all states with 0 counters.
+        // Per-stream synchronized is required to make sure the state (at the head of the queue)
+        // logic is applied to is the state actually removed.
+
         final LinkedList<Long> stateIdQueue = entry.getValue();
         while (true) {
-          final Long oldestState = stateIdQueue.peek();
-          if (oldestState == null) {
+          final Long oldestStateId = stateIdQueue.peek();
+          // no state to flush for this stream
+          if (oldestStateId == null) {
             break;
           }
 
-          // technically possible this map hasn't been updated yet.
-          final boolean noCorrespondingStateMsg = stateIdToState.get(oldestState) == null;
-          if (noCorrespondingStateMsg) {
+          final var oldestState = stateIdToState.get(oldestStateId);
+          // no state to flush for this stream
+          if (oldestState == null) {
             break;
           }
 
-          final boolean noPrevRecs = !stateIdToCounter.containsKey(oldestState);
-          final boolean allRecsEmitted = stateIdToCounter.get(oldestState).get() == 0;
-          if (noPrevRecs || allRecsEmitted) {
-            var polled = entry.getValue().poll(); // poll to remove. no need to read as the earlier peek is still valid.
-            output.add(stateIdToState.get(oldestState).getLeft());
-            bytesFlushed += stateIdToState.get(oldestState).getRight();
+          // technically possible this map hasn't been updated yet.
+          final var oldestStateCounter = stateIdToCounter.get(oldestStateId);
+          Objects.requireNonNull(oldestStateCounter, "Invariant Violation: No record counter found for state message.");
```

Comment:
I think this is only true if we make sure to add a counter before we add the state id to the queue - I think we can do this if line 334 comes before line 333.